### Full Stack Data Science Pro
### Assignment 3 : Python - For loop

# **Name : Aasif Majeed**


## **Basic Level:**

---



## **1. Write a Python program to print the numbers from 1 to 10 using a `for` loop.**

In [ ]:
# Using a for loop to iterate over the range from 1 to 11 (exclusive)
for number in range(1, 11):
    # Printing each number in the loop on the same line
    print(number, end=' ') #end ='' prints numbers in single line reduces code complexity

1 2 3 4 5 6 7 8 9 10 

## **2. Create a program that calculates the sum of all numbers in a list using a `for` loop.**

In [ ]:
# Define a list of numbers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Initialize a variable to store the sum
sum_of_numbers = 0

# Using a for loop to iterate over each number in the list
for number in numbers:
    # Adding each number to the sum
    sum_of_numbers += number # same as sum.. = sum.. + number

# Printing the final sum
print("Sum of numbers:", sum_of_numbers)

Sum of numbers: 55


## **3. Write a program to print the characters of a string in reverse order using a `for` loop.**

In [ ]:
# Define a string
input_string = "Aasif, Majeed"

# Using a for loop to iterate over the characters of the string in reverse order
for char in reversed(input_string):
    # Printing each character in the loop
    print(char, end=' ')

d e e j a M   , f i s a A 

## **4. Develop a program that finds the factorial of a given number using a `for` loop.**

In [ ]:
# Function to calculate the factorial of a number
def calculate_factorial(n):
    # Initialize the factorial variable to 1
    factorial = 1

    # Using a for loop to multiply numbers from 1 to n
    for i in range(1, n + 1):
        # Multiply the current number with the factorial
        factorial *= i

    # Return the final factorial value
    return factorial

# Example: Calculate factorial of 5
number = 5
result = calculate_factorial(number)

# Print the result
print(f"The factorial of {number} is {result}")

The factorial of 5 is 120


## **5. Create a program to print the multiplication table of a given number using a `for` loop.**

In [ ]:
# Function to print the multiplication table of a given number
def multiplication_table(number):
    # Using a for loop to iterate from 1 to 10
    for i in range(1, 11):
        # Calculate the product and print the result
        product = number * i
        print(f"{number} x {i} = {product}")

# Example: Print the multiplication table of 7
given_number = 7
multiplication_table(given_number)

7 x 1 = 7
7 x 2 = 14
7 x 3 = 21
7 x 4 = 28
7 x 5 = 35
7 x 6 = 42
7 x 7 = 49
7 x 8 = 56
7 x 9 = 63
7 x 10 = 70


## **6. Write a program that counts the number of even and odd numbers in a list using a `for` loop.**

In [ ]:
# Define a list of numbers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Initialize variables to count even and odd numbers
even_count = 0
odd_count = 0

# Using a for loop to iterate over each number in the list
for number in numbers:
    # Check if the number is even or odd
    if number % 2 == 0:
        # Increment the even count
        even_count += 1
    else:
        # Increment the odd count
        odd_count += 1

# Printing the counts
print("Number of even numbers:", even_count)
print("Number of odd numbers:", odd_count)

Number of even numbers: 5
Number of odd numbers: 5


## **7. Develop a program that prints the squares of numbers from 1 to 5 using a `for` loop.**

In [ ]:
# Using a for loop to iterate over the range from 1 to 6 (exclusive)
for number in range(1, 6):
    # Calculate the square and print the result
    square = number ** 2
    print(f"The square of {number} is {square}")# "f-string" or "formatted string literal.

The square of 1 is 1
The square of 2 is 4
The square of 3 is 9
The square of 4 is 16
The square of 5 is 25


## **8. Create a program to find the length of a string without using the `len()` function.**

In [ ]:
# Function to find the length of a string without using len()
def find_string_length(input_string):
    # Initialize a counter variable
    length = 0

    # Iterate through each character in the string
    for char in input_string:
        # Increment the counter for each character
        length += 1

    # Return the final length
    return length

# Example: Find the length of a string
input_string = "My name is Aasif Majeed"
result = find_string_length(input_string)

# Print the result
print(f"The length of the string is: {result}")


The length of the string is: 23


## **9. Write a program that calculates the average of a list of numbers using a `for` loop.**

In [ ]:
# Define a list of numbers
numbers = [10, 5, 8, 12, 6]

# Initialize variables for sum and count
total_sum = 0
count = 0

# Using a for loop to iterate over each number in the list
for number in numbers:
    # Add each number to the sum
    total_sum += number
    # Increment the count
    count += 1 #to get the total numbers in the list

# Calculate the average
average = total_sum / count

# Print the result
print(f"The average of the numbers is: {average}")


The average of the numbers is: 8.2


## **10. Develop a program that prints the first `n` Fibonacci numbers using a `for` loop.**

In [ ]:
# Function to print the first n Fibonacci numbers
def fibonacci_sequence(n):
    # Initialize the first two numbers in the sequence
    fib_sequence = [0, 1]

    # Using a for loop to generate the Fibonacci sequence
    for i in range(2, n):
        # Calculate the next Fibonacci number by adding the last two numbers
        next_number = fib_sequence[i - 1] + fib_sequence[i - 2]
        # Append the next number to the sequence
        fib_sequence.append(next_number)

    # Print the Fibonacci sequence
    print(f"The first {n} Fibonacci numbers are: {fib_sequence}")

# Example: Print the first 8 Fibonacci numbers
n = 8
fibonacci_sequence(n)


The first 8 Fibonacci numbers are: [0, 1, 1, 2, 3, 5, 8, 13]


## ***Intermediate Level:***

---



## **11. Write a program to check if a given list contains any duplicates using a `for` loop.**

In [ ]:
# Function to check for duplicates in a list
def has_duplicates(input_list):
    # Create an empty set to store unique elements
    unique_elements = set()

    # Using a for loop to iterate over each element in the list
    for element in input_list:
        # If the element is already in the set, it's a duplicate
        if element in unique_elements:
            return True
        else:
            # Add the element to the set if it's not already present
            unique_elements.add(element)

    # If the loop completes without finding duplicates, return False
    return False

# Example: Check for duplicates in a list
numbers_with_duplicates = [1, 2, 3, 4, 5, 2, 7, 8, 9]
result = has_duplicates(numbers_with_duplicates)

# Print the result
if result:
    print("The list contains duplicates.")
else:
    print("The list does not contain duplicates.")


The list contains duplicates.


## **12. Create a program that prints the prime numbers in a given range using a `for` loop**.

In [ ]:
# Function to check if a number is prime
def is_prime(number):
    # Prime numbers are greater than 1
    if number <= 1:
        return False

    # Check for factors from 2 to the square root of the number
    for i in range(2, int(number**0.5) + 1):
        if number % i == 0:
            return False

    # If no factors were found, the number is prime
    return True

# Function to print prime numbers in a given range
def print_primes_in_range(start, end):
    # Using a for loop to iterate over the range
    for num in range(start, end + 1):
        # Check if the current number is prime
        if is_prime(num):
            print(num, end=' ')

# Example: Print prime numbers in the range 10 to 50
start_range = 10
end_range = 50
print(f"Prime numbers in the range {start_range} to {end_range}:")
print_primes_in_range(start_range, end_range)

Prime numbers in the range 10 to 50:
11 13 17 19 23 29 31 37 41 43 47 

## **13. Develop a program that counts the number of vowels in a string using a `for` loop.**

In [ ]:
# Function to count vowels in a string
def count_vowels(input_string):
    # Define a set of vowels
    vowels = set("aeiouAEIOU")

    # Initialize a counter variable
    vowel_count = 0

    # Using a for loop to iterate over each character in the string
    for char in input_string:
        # Check if the character is a vowel
        if char in vowels:
            vowel_count += 1

    # Return the final count
    return vowel_count

# Example: Count vowels in a string
input_string = "Aasif Majeed"
result = count_vowels(input_string)

# Print the result
print(f"The number of vowels in the string is: {result}")


The number of vowels in the string is: 6


## **14. Write a program to find the maximum element in a 2D list using a nested `for` loop.**

In [ ]:
# Function to find the maximum element in a 2D list
def find_max_element(matrix):
    # Assume the first element as the maximum
    max_element = matrix[0][0]

    # Using nested for loops to iterate through the 2D list
    for row in matrix:
        for element in row:
            # Update max_element if a greater element is found
            if element > max_element:
                max_element = element

    # Return the maximum element
    return max_element

# Example: Find the maximum element in a 2D list
matrix = [
    [3, 8, 1],
    [5, 2, 9],
    [4, 7, 6]
]

result = find_max_element(matrix)

# Print the result
print(f"The maximum element in the 2D list is: {result}")


The maximum element in the 2D list is: 9


## **15. Create a program that removes all occurrences of a specific element from a list using a `for` loop.**

In [ ]:
# Function to remove all occurrences of a specific element from a list
def remove_element(lst, target):
    # Using a for loop to iterate over the list
    for item in lst[:]:  # Iterate over a copy of the list to avoid modifying it during iteration
        # Check if the current item is equal to the target element
        if item == target:
            # Remove the item from the original list
            lst.remove(item)

# Example: Remove all occurrences of 2 from the list
numbers = [1, 2, 3, 2, 4, 2, 5]
target_element = 2

# Print the original list
print("Original List:", numbers)

# Remove all occurrences of the target element
remove_element(numbers, target_element)

# Print the modified list
print(f"List after removing all occurrences of {target_element}:", numbers)


Original List: [1, 2, 3, 2, 4, 2, 5]
List after removing all occurrences of 2: [1, 3, 4, 5]


## **16. Develop a program that generates a multiplication table for numbers from 1 to 5 using a nested `for` loop.**

In [ ]:
# Using nested for loops to generate a multiplication table
for i in range(1, 6):
    for j in range(1, 11):
        # Calculate the product and print the result
        product = i * j
        print(f"{i} x {j} = {product}")

    # Print a newline after each row to separate the tables
    print()


1 x 1 = 1
1 x 2 = 2
1 x 3 = 3
1 x 4 = 4
1 x 5 = 5
1 x 6 = 6
1 x 7 = 7
1 x 8 = 8
1 x 9 = 9
1 x 10 = 10

2 x 1 = 2
2 x 2 = 4
2 x 3 = 6
2 x 4 = 8
2 x 5 = 10
2 x 6 = 12
2 x 7 = 14
2 x 8 = 16
2 x 9 = 18
2 x 10 = 20

3 x 1 = 3
3 x 2 = 6
3 x 3 = 9
3 x 4 = 12
3 x 5 = 15
3 x 6 = 18
3 x 7 = 21
3 x 8 = 24
3 x 9 = 27
3 x 10 = 30

4 x 1 = 4
4 x 2 = 8
4 x 3 = 12
4 x 4 = 16
4 x 5 = 20
4 x 6 = 24
4 x 7 = 28
4 x 8 = 32
4 x 9 = 36
4 x 10 = 40

5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50



## **17. Write a program that converts a list of Fahrenheit temperatures to Celsius using a `for` loop.**

In [ ]:
# Function to convert Fahrenheit to Celsius
def fahrenheit_to_celsius(fahrenheit):
    return (fahrenheit - 32) * 5/9

# List of Fahrenheit temperatures
fahrenheit_temperatures = [32, 68, 86, 104, 212]

# Using a for loop to iterate over each Fahrenheit temperature
for fahrenheit in fahrenheit_temperatures:
    # Convert Fahrenheit to Celsius and print the result
    celsius = fahrenheit_to_celsius(fahrenheit)
    print(f"{fahrenheit} degrees Fahrenheit is {celsius:.2f} degrees Celsius")


32 degrees Fahrenheit is 0.00 degrees Celsius
68 degrees Fahrenheit is 20.00 degrees Celsius
86 degrees Fahrenheit is 30.00 degrees Celsius
104 degrees Fahrenheit is 40.00 degrees Celsius
212 degrees Fahrenheit is 100.00 degrees Celsius


## **18. Create a program to print the common elements from two lists using a `for` loop.**

In [ ]:
# Function to find common elements in two lists
def find_common_elements(list1, list2):
    # Using a for loop to iterate over elements in the first list
    for element1 in list1:
        # Using a nested for loop to compare with elements in the second list
        for element2 in list2:
            # If a common element is found, print it
            if element1 == element2:
                print(f"Common element: {element1}")

# Example: Find common elements in two lists
list_a = [1, 2, 3, 4, 5]
list_b = [3, 4, 5, 6, 7]

print("List A:", list_a)
print("List B:", list_b)

# Call the function to find common elements
find_common_elements(list_a, list_b)


List A: [1, 2, 3, 4, 5]
List B: [3, 4, 5, 6, 7]
Common element: 3
Common element: 4
Common element: 5


## 19. Develop a program that prints the pattern of right-angled triangles using a `for` loop. Use ‘*’ to draw the pattern

In [ ]:
# Function to print a right-angled triangle pattern
def print_right_triangle(rows):
    # Using a for loop to iterate over rows
    for i in range(1, rows + 1):
        # Using a nested for loop to print '*' for each column
        for j in range(1, i + 1):
            print('*', end=' ')
        # Move to the next line after printing each row
        print()

# Example: Print a right-angled triangle pattern with 5 rows
num_rows = 5
print_right_triangle(num_rows)


* 
* * 
* * * 
* * * * 
* * * * * 


## **20. Write a program to find the greatest common divisor (GCD) of two numbers using a `for` loop.**

In [ ]:
# Function to find the greatest common divisor (GCD) of two numbers
def find_gcd(num1, num2):
    # Determine the smaller of the two numbers
    smaller = min(num1, num2)

    # Initialize the GCD variable
    gcd = 1

    # Using a for loop to iterate from 1 to the smaller number
    for i in range(1, smaller + 1):
        # Check if i is a divisor of both numbers
        if num1 % i == 0 and num2 % i == 0:
            # Update the GCD if i is a common divisor
            gcd = i

    # Return the GCD
    return gcd

# Example: Find the GCD of 48 and 18
number1 = 48
number2 = 18

result = find_gcd(number1, number2)

# Print the result
print(f"The GCD of {number1} and {number2} is: {result}")


The GCD of 48 and 18 is: 6


## ***Advanced Level:***

---



## **21. Create a program that calculates the sum of the digits of numbers in a list using a list comprehension.**

In [1]:
# Function to calculate the sum of digits of a number
def sum_of_digits(number):
    return sum(int(digit) for digit in str(number) if digit.isdigit())

# List of numbers
numbers = [123, 456, 789, 12, 34]

# Using a list comprehension to calculate the sum of digits for each number
sums_of_digits = [sum_of_digits(num) for num in numbers]

# Print the result
print("Original Numbers:", numbers)
print("Sums of Digits:", sums_of_digits)


Original Numbers: [123, 456, 789, 12, 34]
Sums of Digits: [6, 15, 24, 3, 7]


## **22. Write a program to find the prime factors of a given number using a `for` loop and list comprehension.**

In [2]:
# Function to find prime factors of a number
def prime_factors(n):
    # Using a list comprehension to find prime factors
    factors = [factor for factor in range(2, n + 1) if n % factor == 0 and all(factor % i != 0 for i in range(2, int(factor**0.5) + 1))]

    return factors

# Example: Find prime factors of 48
number = 48

result = prime_factors(number)

# Print the result
print(f"The prime factors of {number} are: {result}")



The prime factors of 48 are: [2, 3]


simple version of the above problem

In [3]:
# Function to check if a number is prime
def is_prime(num):
    if num < 2:
        return False
    for i in range(2, int(num**0.5) + 1):
        if num % i == 0:
            return False
    return True

# Function to find prime factors of a number
def prime_factors(n):
    factors = []

    # Check for factors from 2 to n
    for factor in range(2, n + 1):
        # If factor is a divisor and prime, add to the list of factors
        if n % factor == 0 and is_prime(factor):
            factors.append(factor)

    return factors

# Example: Find prime factors of 48
number = 48

result = prime_factors(number)

# Print the result
print(f"The prime factors of {number} are: {result}")


The prime factors of 48 are: [2, 3]


## **23. Develop a program that extracts unique elements from a list and stores them in a new list using a list comprehension.**

In [4]:
# Function to extract unique elements from a list using list comprehension
def extract_unique_elements(input_list):
    return list(set(input_list))

# Example: Extract unique elements from a list
original_list = [1, 2, 2, 3, 4, 4, 5, 5, 5]

unique_elements = extract_unique_elements(original_list)

# Print the result
print("Original List:", original_list)
print("Unique Elements:", unique_elements)


Original List: [1, 2, 2, 3, 4, 4, 5, 5, 5]
Unique Elements: [1, 2, 3, 4, 5]


## **24. Create a program that generates a list of all palindromic numbers up to a specified limit using a list comprehension.**

In [5]:
# Function to check if a number is a palindrome
def is_palindrome(num):
    return str(num) == str(num)[::-1]

# Function to generate a list of palindromic numbers up to a specified limit
def generate_palindromic_numbers(limit):
    return [num for num in range(limit) if is_palindrome(num)]

# Example: Generate palindromic numbers up to 1000
limit = 1000

palindromic_numbers = generate_palindromic_numbers(limit)

# Print the result
print(f"Palindromic Numbers up to {limit}: {palindromic_numbers}")


Palindromic Numbers up to 1000: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 22, 33, 44, 55, 66, 77, 88, 99, 101, 111, 121, 131, 141, 151, 161, 171, 181, 191, 202, 212, 222, 232, 242, 252, 262, 272, 282, 292, 303, 313, 323, 333, 343, 353, 363, 373, 383, 393, 404, 414, 424, 434, 444, 454, 464, 474, 484, 494, 505, 515, 525, 535, 545, 555, 565, 575, 585, 595, 606, 616, 626, 636, 646, 656, 666, 676, 686, 696, 707, 717, 727, 737, 747, 757, 767, 777, 787, 797, 808, 818, 828, 838, 848, 858, 868, 878, 888, 898, 909, 919, 929, 939, 949, 959, 969, 979, 989, 999]


## **25. Write a program to flatten a nested list using list comprehension.**

In [6]:
# Function to flatten a nested list using list comprehension
def flatten_nested_list(nested_list):
    return [element for sublist in nested_list for element in sublist]

# Example: Flatten a nested list
nested_list = [[1, 2, 3], [4, 5], [6, 7, 8]]

flattened_list = flatten_nested_list(nested_list)

# Print the result
print("Nested List:", nested_list)
print("Flattened List:", flattened_list)


Nested List: [[1, 2, 3], [4, 5], [6, 7, 8]]
Flattened List: [1, 2, 3, 4, 5, 6, 7, 8]


## **26. Develop a program that computes the sum of even and odd numbers in a list separately using list comprehension.**

In [7]:
# Function to compute the sum of even and odd numbers separately using list comprehension
def sum_even_odd_numbers(numbers):
    even_sum = sum(num for num in numbers if num % 2 == 0)
    odd_sum = sum(num for num in numbers if num % 2 != 0)
    return even_sum, odd_sum

# Example: Compute the sum of even and odd numbers in a list
number_list = [1, 2, 3, 4, 5, 6, 7, 8, 9]

even_sum, odd_sum = sum_even_odd_numbers(number_list)

# Print the result
print("Original List:", number_list)
print("Sum of Even Numbers:", even_sum)
print("Sum of Odd Numbers:", odd_sum)


Original List: [1, 2, 3, 4, 5, 6, 7, 8, 9]
Sum of Even Numbers: 20
Sum of Odd Numbers: 25


## **27. Create a program that generates a list of squares of odd numbers between 1 and 10 using list comprehension.**

In [8]:
# List comprehension to generate squares of odd numbers between 1 and 10
squares_of_odd_numbers = [num**2 for num in range(1, 11) if num % 2 != 0]

# Print the result
print("Squares of Odd Numbers between 1 and 10:", squares_of_odd_numbers)


Squares of Odd Numbers between 1 and 10: [1, 9, 25, 49, 81]


## **28. Write a program that combines two lists into a dictionary using list comprehension.**

In [10]:
# Lists to be combined into a dictionary
keys = ['name', 'age', 'city']
values = ['Aasif', 26, 'Srinagar']

# Using list comprehension to combine lists into a dictionary
combined_dict = {keys[i]: values[i] for i in range(min(len(keys), len(values)))}

# Print the result
print("Combined Dictionary:", combined_dict)


Combined Dictionary: {'name': 'Aasif', 'age': 26, 'city': 'Srinagar'}


## **29. Develop a program that extracts the vowels from a string and stores them in a list using list comprehension.**

In [11]:
# Function to check if a character is a vowel
def is_vowel(char):
    return char.lower() in 'aeiou'

# Function to extract vowels from a string using list comprehension
def extract_vowels(input_string):
    return [char for char in input_string if is_vowel(char)]

# Example: Extract vowels from a string
input_string = "My name is aasif majeed"

vowels_list = extract_vowels(input_string)

# Print the result
print("Original String:", input_string)
print("Vowels Extracted:", vowels_list)


Original String: My name is aasif majeed
Vowels Extracted: ['a', 'e', 'i', 'a', 'a', 'i', 'a', 'e', 'e']


## **30. Create a program that removes all non-numeric characters from a list of strings using list comprehension.**

In [12]:
# Function to filter numeric characters from a string
def filter_numeric_characters(input_str):
    return ''.join(char for char in input_str if char.isdigit())

# Function to remove non-numeric characters from a list of strings using list comprehension
def remove_non_numeric_characters(string_list):
    return [filter_numeric_characters(string) for string in string_list]

# Example: Remove non-numeric characters from a list of strings
string_list = ['abc123', '456def', '7ghi', 'jkl']

numeric_strings = remove_non_numeric_characters(string_list)

# Print the result
print("Original List:", string_list)
print("Numeric Strings:", numeric_strings)



Original List: ['abc123', '456def', '7ghi', 'jkl']
Numeric Strings: ['123', '456', '7', '']


# *Challenge Level:*

## **31. Write a program to generate a list of prime numbers using the Sieve of Eratosthenes algorithm and list comprehension.**

# The Sieve of Eratosthenes is an ancient algorithm for finding all prime numbers up to a given limit. Here's an explanation of how the algorithm works, along with an example:

Create a List: Create a list of consecutive integers from 2 through the given limit. This list is often referred to as the "sieve."

Mark Multiples as Composite: Start with the first prime number, which is 2. Mark all multiples of 2 as composite (not prime). Move to the next unmarked number (which is a prime), and repeat the process until you've processed all numbers up to the square root of the given limit.

Collect Primes: All unmarked numbers in the list are now prime. Collect them into a separate list.

In [15]:
# Function to generate a list of prime numbers using the Sieve of Eratosthenes
def sieve_of_eratosthenes(limit):
    # Create a boolean list representing whether each number is prime
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False  # 0 and 1 are not prime

    # Mark multiples of primes as non-prime
    for i in range(2, int(limit**0.5) + 1):
        if sieve[i]:
            # Use list slicing to mark multiples as non-prime
            sieve[i*i:limit+1:i] = [False] * len(range(i*i, limit+1, i))

    # Return a list of prime numbers by filtering the sieve
    return [num for num in range(2, limit + 1) if sieve[num]]

# Example: Generate a list of prime numbers up to 50
limit = 50

# Call the sieve_of_eratosthenes function to get prime numbers
prime_numbers = sieve_of_eratosthenes(limit)

# Print the result
print(f"Prime Numbers up to {limit}: {prime_numbers}")



Prime Numbers up to 50: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]


In [14]:
#using list comprehension
def sieve_of_eratosthenes(limit):
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False

    [sieve[i*i:limit+1:i] for i in range(2, int(limit**0.5) + 1) if sieve[i]]

    return [num for num in range(2, limit + 1) if sieve[num]]

# Example: Generate a list of prime numbers up to 50
limit = 50

prime_numbers = sieve_of_eratosthenes(limit)

# Print the result
print(f"Prime Numbers up to {limit}: {prime_numbers}")


Prime Numbers up to 50: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


## **32. Create a program that generates a list of all Pythagorean triplets up to a specified limit using list comprehension.**

A Pythagorean triplet is a set of three positive integers
(a, b, c) that satisfy the Pythagorean theorem:

a^2 + b^2 = c^2

In [17]:
# Function to check if a set of three numbers forms a Pythagorean triplet
def is_pythagorean_triplet(a, b, c):
    return a**2 + b**2 == c**2

# Function to generate a list of Pythagorean triplets up to a specified limit
def generate_pythagorean_triplets(limit):
    # Use list comprehension to generate triplets
    triplets = [(a, b, c) for a in range(1, limit + 1)
                              for b in range(a, limit + 1)
                              for c in range(b, limit + 1)
                              if is_pythagorean_triplet(a, b, c)]
    return triplets

# Example: Generate Pythagorean triplets up to a limit of 20
limit = 20

# Generate Pythagorean triplets
pythagorean_triplets = generate_pythagorean_triplets(limit)

# Print the result
print(f"Pythagorean Triplets up to {limit}: {pythagorean_triplets}")





Pythagorean Triplets up to 20: [(3, 4, 5), (5, 12, 13), (6, 8, 10), (8, 15, 17), (9, 12, 15), (12, 16, 20)]


## **33. Develop a program that generates a list of all possible combinations of two lists using list comprehension.**

In [18]:
# Function to generate all possible combinations of two lists
def generate_combinations(list1, list2):
    # Using list comprehension to generate combinations
    combinations = [(x, y) for x in list1 for y in list2]
    return combinations

# Example lists
list_a = [1, 2, 3]
list_b = ['a', 'b', 'c']

# Generate all possible combinations
all_combinations = generate_combinations(list_a, list_b)

# Print the result
print(f"All Combinations of {list_a} and {list_b}: {all_combinations}")


All Combinations of [1, 2, 3] and ['a', 'b', 'c']: [(1, 'a'), (1, 'b'), (1, 'c'), (2, 'a'), (2, 'b'), (2, 'c'), (3, 'a'), (3, 'b'), (3, 'c')]


## **34. Write a program that calculates the mean, median, and mode of a list of numbers using list comprehension.**

In [19]:
from collections import Counter

# Function to calculate mean of a list
def calculate_mean(numbers):
    return sum(numbers) / len(numbers) if len(numbers) > 0 else 0

# Function to calculate median of a list
def calculate_median(numbers):
    sorted_numbers = sorted(numbers)
    n = len(sorted_numbers)
    mid = n // 2
    return (sorted_numbers[mid] + sorted_numbers[mid - 1]) / 2 if n % 2 == 0 else sorted_numbers[mid]

# Function to calculate mode of a list
def calculate_mode(numbers):
    counts = Counter(numbers)
    mode = max(counts, key=counts.get)
    return mode

# Example list of numbers
number_list = [1, 2, 3, 4, 5, 5, 6, 7, 8, 8, 9]

# Calculate mean, median, and mode
mean_value = calculate_mean(number_list)
median_value = calculate_median(number_list)
mode_value = calculate_mode(number_list)

# Print the results
print(f"List of Numbers: {number_list}")
print(f"Mean: {mean_value}")
print(f"Median: {median_value}")
print(f"Mode: {mode_value}")


List of Numbers: [1, 2, 3, 4, 5, 5, 6, 7, 8, 8, 9]
Mean: 5.2727272727272725
Median: 5
Mode: 5


## **35. Create a program that generates Pascal's triangle up to a specified number of rows using list comprehension.**

# Pascal's Triangle

Pascal's Triangle is a mathematical structure named after the French mathematician Blaise Pascal, although it was known to Chinese mathematicians over 500 years earlier.

It is formed by starting with the number 1 at the apex, and then constructing the rows below it, with each entry being the sum of the two numbers immediately above it. The triangle is infinite, but usually, a specified number of rows are generated.

## Construction

The first few rows of Pascal's Triangle look like this:



      1
     1 1
    1 2 1


This row represents the coefficients of the binomial expansion of (a + b)^3, where the exponents are 3, 2, 1, and 0, respectively.

In general, the nth row of Pascal's Triangle represents the coefficients of the binomial expansion of (a + b)^(n-1).

Pascal's Triangle has wide applications in combinatorics, probability theory, and algebra.

In [21]:
# Function to generate Pascal's triangle up to a specified number of rows
def generate_pascals_triangle(rows):
    triangle = [[1] * (row + 1) for row in range(rows)]
    for row in range(2, rows):
        for col in range(1, row):
            triangle[row][col] = triangle[row - 1][col - 1] + triangle[row - 1][col]
    return triangle

# Function to display Pascal's triangle
def display_pascals_triangle(triangle):
    for row in triangle:
        print(row)

# Example: Generate Pascal's triangle with 5 rows
num_rows = 5

# Generate Pascal's triangle
pascals_triangle = generate_pascals_triangle(num_rows)

# Display the result
print(f"Pascal's Triangle (up to {num_rows} rows):")
display_pascals_triangle(pascals_triangle)



Pascal's Triangle (up to 5 rows):
[1]
[1, 1]
[1, 2, 1]
[1, 3, 3, 1]
[1, 4, 6, 4, 1]


## **36. Develop a program that calculates the sum of the digits of a factorial of numbers from 1 to 5 using list comprehension.**

In [23]:
import math

# Function to calculate the sum of digits of a number
def sum_of_digits(number):
    return sum(int(digit) for digit in str(number))

# Calculate the factorial of numbers from 1 to 5 and their sum of digits using list comprehension
factorials_and_sums = [(n, sum_of_digits(math.factorial(n))) for n in range(1, 6)]

# Print the results
for n, digit_sum in factorials_and_sums:
    print(f"Factorial of {n}! is {math.factorial(n)}, and the sum of its digits is {digit_sum}")


Factorial of 1! is 1, and the sum of its digits is 1
Factorial of 2! is 2, and the sum of its digits is 2
Factorial of 3! is 6, and the sum of its digits is 6
Factorial of 4! is 24, and the sum of its digits is 6
Factorial of 5! is 120, and the sum of its digits is 3


## **37. Write a program that finds the longest word in a sentence using list comprehension.**

In [25]:
# Function to find the longest word in a sentence
def find_longest_word(sentence):
    words = sentence.split()
    longest_word = max(words, key=len)
    return longest_word

# Example sentence
example_sentence = "My name is AASIFMAJEED, and i am from kashmir"

# Find the longest word using list comprehension
longest_word = find_longest_word(example_sentence)

# Print the result
print(f"The longest word in the sentence is: {longest_word}")


The longest word in the sentence is: AASIFMAJEED,


## **38. Create a program that filters a list of strings to include only those with more than three vowels using list comprehension.**

In [29]:
# Function to count vowels in a string
def count_vowels(string):
    return sum(1 for char in string if char.lower() in 'aeiou')

# Function to filter strings with more than three vowels
def filter_strings_by_vowels(string_list):
    return [s for s in string_list if count_vowels(s) > 3]

# Example list of strings
example_strings = ["hello", "world", "python", "programming", "example","aasiif"]

# Filter strings with more than three vowels
filtered_strings = filter_strings_by_vowels(example_strings)

# Print the result
print(f"Original List: {example_strings}")
print(f"Filtered List with more than three vowels: {filtered_strings}")


Original List: ['hello', 'world', 'python', 'programming', 'example', 'aasiif']
Filtered List with more than three vowels: ['aasiif']


## **39. Develop a program that calculates the sum of the digits of numbers from 1 to 1000 using list comprehension.**

In [30]:
# Function to calculate the sum of digits of a number
def sum_of_digits(number):
    return sum(int(digit) for digit in str(number))

# Calculate the sum of digits for numbers from 1 to 1000 using list comprehension
sums_of_digits = [sum_of_digits(n) for n in range(1, 1001)]

# Print the result
total_sum = sum(sums_of_digits)
print(f"The sum of digits for numbers from 1 to 1000 is: {total_sum}")


The sum of digits for numbers from 1 to 1000 is: 13501


## **40. Write a program that generates a list of prime palindromic numbers using list comprehension.**

# Prime Palindromic Numbers

Prime Palindromic Numbers are special numbers that possess two distinct properties: they are both prime and palindromic.

## Prime Numbers

A prime number is a natural number greater than 1 that is not a product of two smaller natural numbers. In other words, a prime number has no positive divisors other than 1 and itself. Examples of prime numbers include 2, 3, 5, 7, 11, and so on.

## Palindromic Numbers

A palindromic number is a number that remains the same when its digits are reversed. For example, 121, 1331, and 1221 are palindromic numbers.

## Prime Palindromic Numbers

Prime Palindromic Numbers are those rare numbers that satisfy both conditions—they are prime and palindromic simultaneously. Generating and exploring these numbers can be an interesting mathematical exercise.




In [31]:
# Function to check if a number is prime
def is_prime(number):
    if number < 2:
        return False
    for i in range(2, int(number**0.5) + 1):
        if number % i == 0:
            return False
    return True

# Function to check if a number is palindromic
def is_palindromic(number):
    return str(number) == str(number)[::-1]

# Generate a list of prime palindromic numbers using list comprehension
prime_palindromes = [num for num in range(10, 1001) if is_prime(num) and is_palindromic(num)]

# Print the result
print("Prime Palindromic Numbers:", prime_palindromes)


Prime Palindromic Numbers: [11, 101, 131, 151, 181, 191, 313, 353, 373, 383, 727, 757, 787, 797, 919, 929]


## Program Explanation

The provided Python program uses list comprehension to generate a list of Prime Palindromic Numbers in a specified range (in this case, between 10 and 1000). The program employs two functions: one to check if a number is prime (`is_prime`), and the other to check if a num